src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -319,5 +321,28 @@ protected override void Dispose(bool disposing)
 
             public CancellationTokenRegistration CancellationTokenRegistration { get; }
         }
+
+        private class WrappedTCS : TaskCompletionSource<RateLimitLease>
+        {
+            private readonly int PermitCount;
+            private readonly ConcurrencyLimiter Limiter;
+
+            public WrappedTCS(int permitCount, ConcurrencyLimiter limiter) : base(TaskCreationOptions.RunContinuationsAsynchronously)
+            {
+                PermitCount = permitCount;
+                Limiter = limiter;
+            }
+
+            public void TryCancel()
+            {
+                if (TrySetException(new OperationCanceledException()))
```

Comment:
Why this rather than `TrySetCanceled()`?

Reply:
Here's some previous discussion: https://github.com/aspnet/AspLabs/pull/387#discussion_r720557517

Given that we're doing `return new ValueTask<RateLimitLease>(Task.FromCanceled<RateLimitLease>(cancellationToken));` instead of calling `ThrowIfCancellationRequested()`, that's even more reason to use `TrySetCanceled()` now here.

We should probably capture the `cancellationToken` too and call the `TrySetCanceled(CancellationToken)` overload.